manager-rs/dauth-service/src/database/flood_vectors.rs

```diff
@@ -0,0 +1,422 @@
+use sqlx::sqlite::{SqlitePool, SqliteRow};
+use sqlx::Error as SqlxError;
+use sqlx::{Sqlite, Transaction};
+
+use crate::data::error::DauthError;
+
+/// Creates the flood vector table if it does not exist already.
+/// Meant to be used before the auth vector table.
+pub async fn init_table(pool: &SqlitePool) -> Result<(), DauthError> {
+    sqlx::query(
+        "CREATE TABLE IF NOT EXISTS flood_vector_table (
+            user_id TEXT NOT NULL,
+            seqnum INT NOT NULL,
+            xres_star_hash BLOB NOT NULL,
+            autn BLOB NOT NULL,
+            rand BLOB NOT NULL,
+            PRIMARY KEY (user_id, seqnum)
+        );",
+    )
+    .execute(pool)
+    .await?;
+    Ok(())
+}
+
+/// Inserts a vector with the given data.
+/// Returns an error if (id, seqnum) is not unique.
+pub async fn add(
+    transaction: &mut Transaction<'_, Sqlite>,
+    id: &str,
+    seqnum: i64,
+    xres_star_hash: &[u8],
+    autn: &[u8],
+    rand: &[u8],
+) -> Result<(), DauthError> {
+    sqlx::query(
+        "INSERT INTO flood_vector_table
+        VALUES ($1,$2,$3,$4,$5)",
+    )
+    .bind(id)
+    .bind(seqnum)
+    .bind(xres_star_hash)
+    .bind(autn)
+    .bind(rand)
+    .execute(transaction)
+    .await?;
+
+    Ok(())
+}
+
+/// Check first for higher priority flood vectors
+/// Generally, it is normal for this to return Ok(None)
+pub async fn get_first(
+    transaction: &mut Transaction<'_, Sqlite>,
+    id: &str,
+) -> Result<Option<SqliteRow>, SqlxError> {
+    let res = sqlx::query(
+        "SELECT * FROM flood_vector_table
+        WHERE user_id=$1
+        ORDER BY seqnum
+        LIMIT 1;",
+    )
+    .bind(id)
+    .fetch_one(transaction)
+    .await;
+
+    match res {
+        Err(SqlxError::RowNotFound) => Ok(None),
+        _ => Ok(Some(res?)),
+    }
+}
+
+/// Removes the vector with the (id, seqnum) pair.
+pub async fn remove(
+    transaction: &mut Transaction<'_, Sqlite>,
+    id: &str,
+    seqnum: i64,
+) -> Result<(), DauthError> {
+    sqlx::query(
+        "DELETE FROM flood_vector_table
+        WHERE (user_id,seqnum)=($1,$2)",
+    )
+    .bind(id)
+    .bind(seqnum)
+    .execute(transaction)
+    .await?;
+    Ok(())
+}
+
+/// Removes all vectors belonging to an id.
+pub async fn _remove_all(
```

Comment:
Ah, that was an attempt to avoid the compiler warning for unused functions... but there are quite a few since this is a work in progress, so I'll remove it